### PR TITLE
Fix missing dagruns when catchup=True

### LINKED ISSUE #19461: Missing DagRuns when catchup=True
### Apache Airflow version

2.2.1 (latest released)

### Operating System

PRETTY_NAME="Debian GNU/Linux 10 (buster)

### Versions of Apache Airflow Providers

apache-airflow-providers-ftp==2.0.1
apache-airflow-providers-http==2.0.1
apache-airflow-providers-imap==2.0.1
apache-airflow-providers-sqlite==2.0.1

### Deployment

Docker-Compose

### Deployment details

_No response_

### What happened

Backfilling via catchup=True leads to missing DagRuns. 

See reproduction steps for full details

### What you expected to happen

_No response_

### How to reproduce

Note, this is an issue which we have experienced in our production environment, with a much more complicated DAG. Below are the reproduction steps using breeze.

1. Setup ./breeze environment with the below config modifications
2. Create a simple DAG, with dummy tasks in it (see below example)
2. Set a `start_date` in the past
3. Set `catchup=True`
4. Unpause the DAG
5. Catch up starts and if you view the tree view, you have the false impression that everything has caught up correctly. 
6. However, access the calendar view, you can see the missing DagRuns. 

**Breeze Config**
```
export DB_RESET="true"
export START_AIRFLOW="true"
export INSTALL_AIRFLOW_VERSION="2.2.1"
export USE_AIRFLOW_VERSION="2.2.1"
```

**Dummy DAG**
```
from datetime import datetime
from airflow import DAG
from airflow.operators.dummy import DummyOperator

dag = DAG(
    dag_id="temp_test",
    schedule_interval="@daily",
    catchup=True,
    start_date=datetime(2021, 8, 1),
    max_active_tasks=10,
    max_active_runs=5,
    is_paused_upon_creation=True,
)

with dag:
    task1 = DummyOperator(task_id="task1")
    task2 = DummyOperator(task_id="task2")
    task3 = DummyOperator(task_id="task3")
    task4 = DummyOperator(task_id="task4")
    task5 = DummyOperator(task_id="task5")

    task1 >> task2 >> task3 >> task4 >> task5
```

**Results**
tree_view

calendar_view

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by ephraimbuddy**
Closes: https://github.com/apache/airflow/issues/19461
There's a bug that when the max_active_runs is reached, run dates could skip.
This PR fixes it

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/jobs/scheduler_job.py [hunk 1]
```diff
@@ -913,7 +913,8 @@ def _create_dag_runs(self, dag_models: Collection[DagModel], session: Session) -
                     creating_job_id=self.id,
                 )
                 active_runs_of_dags[dag.dag_id] += 1
-            self._update_dag_next_dagruns(dag, dag_model, active_runs_of_dags[dag.dag_id])
+            if self._update_dag_next_dagruns(dag, dag_model, active_runs_of_dags[dag.dag_id]):
+                dag_model.calculate_dagrun_date_fields(dag, data_interval)
```
Comment (jedcunningham) [inline airflow/jobs/scheduler_job.py]: With this change I think the name `_update_dag_next_dagruns` is misleading. Either rename to `_should_update_dag_next_dagruns`, or keep calling `calculate_dagrun_date_fields` and get the interval from `dag.get_run_data_interval(dag_run)`?

File: tests/jobs/test_scheduler_job.py [hunk 2]
```diff
@@ -2874,6 +2875,66 @@ def test_more_runs_are_not_created_when_max_active_runs_is_reached(self, dag_mak
             == 0
         )
 
+    def test_max_active_runs_creation_phasing(self, dag_maker, session):
+        """
+        Test that when creating runs once max_active_runs is reached that the dags come in the right order
```
Comment (jedcunningham) [inline tests/jobs/test_scheduler_job.py]:
```suggestion
        Test that when creating runs once max_active_runs is reached that the runs come in the right order
```

File: tests/jobs/test_scheduler_job.py [hunk 3]
```diff
@@ -2874,6 +2875,66 @@ def test_more_runs_are_not_created_when_max_active_runs_is_reached(self, dag_mak
...
+                2016,
+                1,
+                3,
+            )
+        )
+        assert model.next_dagrun_create_after is None
+
+        complete_one_dagrun()
+
+        assert DagRun.active_runs_of_dags(session=session) == {'test_dag': 3}, "Test only. XXX Remove me"
```
Comment (jedcunningham) [inline tests/jobs/test_scheduler_job.py]:
```suggestion
        assert DagRun.active_runs_of_dags(session=session) == {'test_dag': 3}
```

File: tests/jobs/test_scheduler_job.py [hunk 4]
```diff
@@ -1274,7 +1274,8 @@ def test_dagrun_timeout_fails_run_and_update_next_dagrun(self, dag_maker):
         session.flush()
         session.refresh(dr)
         assert dr.state == State.FAILED
-        # check that next_dagrun has been updated by Schedulerjob._update_dag_next_dagruns
+        self.scheduler_job._do_scheduling(session)
```
Comment (ephraimbuddy) [inline tests/jobs/test_scheduler_job.py]: I'm not sure if this test is correct or not, the assert below was failing, that was why I made the change here. Could be another bug, please review

File: airflow/jobs/scheduler_job.py (hunk 1 above)
Comment (ephraimbuddy) [reply airflow/jobs/scheduler_job.py]: I will go with renaming cause calling `calculate_dagrun_date_fields` only won't handle the `max_active_runs` for the dag

File: airflow/models/dag.py [hunk 5]
```diff
@@ -630,6 +630,7 @@ def get_next_data_interval(self, dag_model: "DagModel") -> Optional[DataInterval
         data_interval = dag_model.next_dagrun_data_interval
         if data_interval is not None:
             return data_interval
+        self.log.warning("%s needed to infer interval", self.dag_id)
```
Comment (ashb) [inline airflow/models/dag.py]: This was just my debugging/testing and we don't need this here I don't think.

Also it can stay if we think it's useful

File: airflow/jobs/scheduler_job.py [hunk 6]
```diff
@@ -913,26 +913,26 @@ def _create_dag_runs(self, dag_models: Collection[DagModel], session: Session) -
...
                 )
                 active_runs_of_dags[dag.dag_id] += 1
-            self._update_dag_next_dagruns(dag, dag_model, active_runs_of_dags[dag.dag_id])
+            if self._should_update_dag_next_dagruns(dag, dag_model, active_runs_of_dags[dag.dag_id]):
+                dag_model.calculate_dagrun_date_fields(dag, data_interval)
         # TODO[HA]: Should we do a session.flush() so we don't have to keep lots of state/object in
         # memory for larger dags? or expunge_all()
 
-    def _update_dag_next_dagruns(self, dag, dag_model: DagModel, total_active_runs) -> None:
+    def _should_update_dag_next_dagruns(self, dag, dag_model: DagModel, total_active_runs) -> None:
```
Comment (jedcunningham) [inline airflow/jobs/scheduler_job.py]:
```suggestion
    def _should_update_dag_next_dagruns(self, dag, dag_model: DagModel, total_active_runs) -> bool:
```

Comment (kaxil) [reply airflow/jobs/scheduler_job.py]: Needs updating docstrings too in that case

File: airflow/models/dag.py (hunk 5 above)
Comment (ephraimbuddy) [reply airflow/models/dag.py]: Happy to remove it then!

Comment (kaxil) [reply airflow/models/dag.py]: The message does not make sense.

"'example_bash_operator' needed to infer interval"

```suggestion
```

Comment (kaxil) [reply airflow/models/dag.py]: If however, we want to add it, we probably add it in `infer_automated_data_interval`, something like:

"Inferring Data Interval for self.dag_id"

File: tests/jobs/test_scheduler_job.py (hunk 3 above)
Comment (kaxil) [reply tests/jobs/test_scheduler_job.py]: @ephraimbuddy  ^^

File: airflow/jobs/scheduler_job.py [hunk 7]
```diff
@@ -913,26 +913,23 @@ def _create_dag_runs(self, dag_models: Collection[DagModel], session: Session) -
                     creating_job_id=self.id,
                 )
                 active_runs_of_dags[dag.dag_id] += 1
-            self._update_dag_next_dagruns(dag, dag_model, active_runs_of_dags[dag.dag_id])
+            if self._should_update_dag_next_dagruns(dag, dag_model, active_runs_of_dags[dag.dag_id]):
```
Comment (jedcunningham) [inline airflow/jobs/scheduler_job.py:916]: I still wonder if having `data_interval` as an optional kwarg to `_update_dag_next_dagruns` would keep this cleaner. Then this case can pass it, the others that don't can get it from `get_run_data_interval`, and we don't have to add an extra conditional and call to `calculate_dagrun_date_fields` in more than 1 place.

Comment (jedcunningham) [reply airflow/jobs/scheduler_job.py:916]: e.g:
```
--- a/airflow/jobs/scheduler_job.py
+++ b/airflow/jobs/scheduler_job.py
@@ -913,11 +913,11 @@ class SchedulerJob(BaseJob):
                     creating_job_id=self.id,
                 )
                 active_runs_of_dags[dag.dag_id] += 1
-            self._update_dag_next_dagruns(dag, dag_model, active_runs_of_dags[dag.dag_id])
+            self._update_dag_next_dagruns(dag, dag_model, active_runs_of_dags[dag.dag_id], data_interval)
         # TODO[HA]: Should we do a session.flush() so we don't have to keep lots of state/object in
         # memory for larger dags? or expunge_all()
 
-    def _update_dag_next_dagruns(self, dag, dag_model: DagModel, total_active_runs) -> None:
+    def _update_dag_next_dagruns(self, dag, dag_model: DagModel, total_active_runs, data_interval = None) -> None:
         """
         Update the next_dagrun, next_dagrun_data_interval_start/end
         and next_dagrun_create_after for this dag.
@@ -931,7 +931,8 @@ class SchedulerJob(BaseJob):
             )
             dag_model.next_dagrun_create_after = None
         else:
-            data_interval = dag.get_next_data_interval(dag_model)
+            if not data_interval:
+                data_interval = dag.get_next_data_interval(dag_model)
             dag_model.calculate_dagrun_date_fields(dag, data_interval)
```

Probably safer with a sentinel instead of None, but I think it'd be okay.

Comment (jedcunningham) [reply airflow/jobs/scheduler_job.py:916]: (I don't have a strong preference here though)

File: tests/jobs/test_scheduler_job.py [hunk 8]
```diff
@@ -2874,6 +2874,66 @@ def test_more_runs_are_not_created_when_max_active_runs_is_reached(self, dag_mak
...
+        assert model.next_dagrun_create_after is None
+
+        complete_one_dagrun()
+
+        assert DagRun.active_runs_of_dags(session=session) == {'test_dag': 3}
+
+        for _ in range(5):
+            self.scheduler_job._do_scheduling(session)
+            complete_one_dagrun()
+            print(session.query(DagRun).order_by(DagRun.execution_date).all())
```
Comment (kaxil) [inline tests/jobs/test_scheduler_job.py]:
```suggestion
```

**Review by kaxil — APPROVED** [review]

File: airflow/jobs/scheduler_job.py (hunk 7 above)
Comment (kaxil) [reply airflow/jobs/scheduler_job.py:916]: No strong preference either but I might merge it tomorrow to cut rc2 and we can refactor if need be after it

Comment (ephraimbuddy) [reply airflow/jobs/scheduler_job.py:916]: This will still have the missing dagruns issue because it's calling `dag.get_next_data_interval` when a run has finished. This will skip it. Using the latest dagrun when the run has just finished and calling `dag.get_run_data_interval` resolves the issue. The `get_next_data_interval` gives us the `next` dagrun's data interval but `get_run_data_interval` when supplied with a recent dagrun gives a dagruns data interval that sets the DagRunInfo correctly especially the `next_dagrun_create_after` for the dag. That way, the nullified `dag_model.next_dagrun_create_after` is set again properly.

File: airflow/jobs/scheduler_job.py [hunk 9]
```diff
@@ -913,26 +913,23 @@ def _create_dag_runs(self, dag_models: Collection[DagModel], session: Session) -
...
-        if total_active_runs >= dag_model.max_active_runs:
+    def _should_update_dag_next_dagruns(self, dag, dag_model: DagModel, total_active_runs) -> bool:
+        """Check if the dag's next_dagruns_create_after should be updated."""
+        if total_active_runs >= dag.max_active_runs:
             self.log.info(
                 "DAG %s is at (or above) max_active_runs (%d of %d), not creating any more runs",
                 dag_model.dag_id,
                 total_active_runs,
-                dag_model.max_active_runs,
+                dag.max_active_runs,
```
Comment (uranusjr) [inline airflow/jobs/scheduler_job.py:928]: I seem to recall there’s another issue on why `max_active_runs` (or something else) does not agree on `DAG` and `DagModel`. Is this change related to that? Did we ever figure out why those disagree?

Comment (ephraimbuddy) [reply airflow/jobs/scheduler_job.py:928]: It's been worked on at https://github.com/apache/airflow/pull/19367. I just decided to update it here as the other PR deals with some other things and is not included in this release

**Review by uranusjr — APPROVED** [review]

**Comment by massamany** [comment]
Hi,

I'm not 100% sure, but I think that I'm experiencing a bug related to this issue.

I currently use Airflow 2.2.0 to schedule several DAGs. I've one which is scheduled to run every 2 minutes.
In this DAG, I'm using the latest_only operator with the catchup activated.
The start date of the DAG is computed before launching Airflow and is valued with the start date of the previous scheduling period.

Description of my reproduction case :
- At startup, the database is empty.
- The DAG is paused at first launch by default.
- After Airflow is up, I wait several minutes and then activate the DAG.


In 2.2.0 :
- Airflow up at 11:11
- Start date computed to be 11:08 as expected
- when the DAG is activated at 11:18, the missed execution are launched, but skipped thanks to the latest_only.
- the last execution runs OK.
- then the DAG is correctly scheduled every 3 minutes.
- conclusion : everything is OK

See screeshots of the DAG details :
image

See screenshots of the Tree view after several minutes (last execution at 11:22 for the scheduling period [11:20-11:22]) :
image


In 2.2.1 AND 2.2.2 :
- Airflow up at 11:24
- Start date computed to be 11:22 as expected
- when the DAG is activated at 11:32, the first missed execution (for 11:22) is launched, but skipped thanks to the latest_only.
- then 2 minutes later (at 11:34), the second execution (for 11:24) is launched, but skipped thanks to the latest_only.
- then the DAG is stuck in this state, the "next run" field still indicates 11:24 and no execution will never be scheduled anymore.
- conclusion : may-be I'm doing something wrong but it seems to me that there might be a bug here related to this current issue.

See screeshots of the DAG details :
image

See screenshots of the full Tree view screen after several minutes (taken at 11:41) :
image

**Comment by ephraimbuddy** [comment]
Dynamic start_date is not good and can give some unexpected behaviours. Can you change to using a static start_date and see if the issue still occurs?

**Comment by massamany** [comment]
Hi. The start date is computed before startup and will never change after.
In fact the DAG python file is generated and start date is computed by generation.
So we can consider it's static from the point of view of Airflow.

**Comment by massamany** [comment]
To make it all simpler, I reproduced the problem with the very simple DAG below.
The start date is static and should be changed to be for instance 10 minutes before startup.
Everything goes on well on Airflow 2.2.0, the scheduling is blocked with 2.2.2.

test_dag.py :
```
import dateutil.parser
from airflow import DAG
from airflow.settings import TIMEZONE
from airflow.operators.latest_only import LatestOnlyOperator
from airflow.operators.python import PythonOperator


def create_test_dag():
    start_date = dateutil.parser.isoparse("2021-11-24T15:00:00+01:00").astimezone(tz=TIMEZONE)

    my_test_dag = DAG(
        dag_id="test_dag",
        start_date=start_date,
        schedule_interval="0/2 * * * *",
        is_paused_upon_creation=True,
        catchup=True,
        default_args={'owner': 'airflow', 'start_date': start_date},
    )

    latest_task = LatestOnlyOperator(task_id="test_latest_only", dag=my_test_dag)
    test_task = PythonOperator(task_id="test_task", dag=my_test_dag, python_callable=lambda: print("Hello guys !"))

    latest_task >> test_task

    return my_test_dag


my_test_dag = create_test_dag()
```

**Comment by massamany** [comment]
If you need the airflow config file I can provide it.
